pkg/kv/kvserver/store_create_replica.go

```diff
@@ -106,64 +167,40 @@ func (s *Store) tryGetOrCreateReplica(
-	if repl, ok := s.mu.replicasByRangeID.Load(rangeID); ok {
-		repl.raftMu.Lock() // not unlocked on success
-		repl.mu.RLock()
-
-		// The current replica is removed, go back around.
-		if repl.mu.destroyStatus.Removed() {
-			repl.mu.RUnlock()
-			repl.raftMu.Unlock()
-			return nil, false, errRetry
-		}
+	if repl, err := s.tryGetReplica(ctx, rangeID, replicaID, creatingReplica); err != nil {
+		return nil, false, err
+	} else if repl != nil {
+		return repl, false, nil
+	}
 
-		// Drop messages from replicas we know to be too old.
-		if fromReplicaIsTooOldRLocked(repl, creatingReplica) {
-			repl.mu.RUnlock()
-			repl.raftMu.Unlock()
-			return nil, false, roachpb.NewReplicaTooOldError(creatingReplica.ReplicaID)
-		}
+	// No replica currently exists, so we'll try to create one. Multiple
+	// goroutines may be racing at this point, so grab a "lock" over this rangeID
+	// in s.creatingReplicas map for one goroutine, and retry others.
+	s.creatingReplicas.Lock()
+	defer s.creatingReplicas.Unlock()
+	if _, ok := s.creatingReplicas.m[rangeID]; ok {
+		return nil, false, errRetry
+	}
+	s.creatingReplicas.m[rangeID] = struct{}{}
+	defer delete(s.creatingReplicas.m, rangeID)
+	// Now we are the only goroutine trying to create a replica for this rangeID.
 
-		// The current replica needs to be removed, remove it and go back around.
-		if toTooOld := repl.replicaID < replicaID; toTooOld {
-			if shouldLog := log.V(1); shouldLog {
-				log.Infof(ctx, "found message for replica ID %d which is newer than %v",
-					replicaID, repl)
-			}
-
-			repl.mu.RUnlock()
-			if err := s.removeReplicaRaftMuLocked(ctx, repl, replicaID, RemoveOptions{
-				DestroyData: true,
-			}); err != nil {
-				log.Fatalf(ctx, "failed to remove replica: %v", err)
-			}
-			repl.raftMu.Unlock()
-			return nil, false, errRetry
-		}
-		defer repl.mu.RUnlock()
-
-		if repl.replicaID > replicaID {
-			// The sender is behind and is sending to an old replica.
-			// We could silently drop this message but this way we'll inform the
-			// sender that they may no longer exist.
-			repl.raftMu.Unlock()
-			return nil, false, &roachpb.RaftGroupDeletedError{}
-		}
-		if repl.replicaID != replicaID {
-			// This case should have been caught by handleToReplicaTooOld.
-			log.Fatalf(ctx, "intended replica id %d unexpectedly does not match the current replica %v",
-				replicaID, repl)
-		}
+	// Repeat the quick path in case someone has overtaken us while we were
+	// grabbing the "lock".
+	if repl, err := s.tryGetReplica(ctx, rangeID, replicaID, creatingReplica); err != nil {
+		return nil, false, err
+	} else if repl != nil {
 		return repl, false, nil
 	}
+	// Now we have the guarantee that s.mu.replicasByRangeID does not contain
+	// rangeID, and only us can insert this rangeID. This also implies that the
+	// RangeTombstone in storage for this rangeID is "locked".
+	//
+	// FIXME: revisit callers of addReplicaToRangeMapLocked, we might need to lock
```

Comment:
Tests pass. But still need to revisit.

Reply:
I'm not sold on introducing another mutex here. Why don't we use `s.mu`, i.e. house `creatingReplicas` within? Each additional mutex brings more cognitive overhead and this path is not "hot" relative to other uses of `s.mu`. It would strike me as problematic if the use of this mutex played a role anywhere but in this method.

Btw, `addReplicaToRangeMap` is called from https://github.com/cockroachdb/cockroach/blob/352d2ef4ccc216dbefb58303789c0f5918bdc81b/pkg/kv/kvserver/store_split.go#L345-L347 (via that method in the snippet), do you know how that works? I thought `rightRepl` was an uninitialized replica already at that point, so isn't it already in there? Seems like it because this comment confirms it:

https://github.com/cockroachdb/cockroach/blob/0703807d018fb1c0f5352e8d83315b463c90d1e9/pkg/kv/kvserver/store_create_replica.go#L374-L380

That seems like another source of confusion. We should require that a replica is already in the range map in `addReplicaInternalLocked`. In other word, nobody should be inserting into the rangeID map except this method right here. This would require an update to `(*Store).Start` (which we want to update anyway, ideally) and some test code (need to change `AddReplica` to first call `getOrCreateReplica`). Maybe worth a TODO or issue?

Reply:
Reusing `s.mu` sounds reasonable. I forgot that we are unlocking the overarching mutex promptly, and then it's the map entry which is the actual "lock".

Still digesting the rest of the comment, will address later.

Reply:
All done in #94996.